pkg/apis/app.kubermatic/v1/ApplicationDefinition.go

```diff
@@ -0,0 +1,183 @@
+	// Username holds the ref and key in the secret for the username credential
+	Username GlobalSecretKeySelector `json:"username"`
+
+	// Password holds the ref and key in the secret for the Password credential
+	Password GlobalSecretKeySelector `json:"password"`
+}
+
+type HelmSource struct {
+	// URl of the helm repository
+	Url string `json:"url"`
+
+	// Name of the Chart
+	ChartName string `json:"chartName"`
+
+	// Version of the Chart
+	ChartVersion string `json:"chartVersion"`
+
+	// Credentials hold the ref to the secret with helm credentials
+	Credentials *HelmCredentials `json:"credentials,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=password;token;ssh-key
+type GitAuthMethod string
+
+type GitCredentials struct {
+	// Authentication method
+	Method GitAuthMethod `json:"method"`
+
+	// Username holds the ref and key in the secret for the username credential
+	Username *GlobalSecretKeySelector `json:"username,omitempty"`
+
+	// Password holds the ref and key in the secret for the Password credential
+	Password *GlobalSecretKeySelector `json:"password,omitempty"`
+
+	// Token holds the ref and key in the secret for the token credential
+	Token *GlobalSecretKeySelector `json:"token,omitempty"`
+
+	// SshKey holds the ref and key in the secret for the SshKey credential
+	SshKey *GlobalSecretKeySelector `json:"sshKey,omitempty"`
+}
+
+type GitSource struct {
+	// URL to the repository (e.g. git://host.xz[:port]/path/to/repo.git/)
+	Remote string `json:"remote"`
+
+	// Git reference to check out: can be a branch, tag, or sha1
+	Ref string `json:"ref"`
+
+	// Path of the "source" in the repository. default is repository root
+	Path string `json:"path,omitempty"`
+
+	// Credentials holds the git credentials
+	Credentials *GitCredentials `json:"credentials,omitempty"`
+}
+
+type ApplicationSource struct {
+	Helm *HelmSource `json:"helm,omitempty"`
+	Git  *GitSource  `json:"git,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=number;boolean;text;text-area
+type FormFieldType string
+
+type FormField struct {
+	// Type of displayed control
+	Type FormFieldType `json:"type"`
+
+	// DisplayName is visible in the UI
+	DisplayName string `json:"displayName"`
+
+	// InternalName is used internally to save in the ApplicationInstallation object
+	InternalName string `json:"internalName"`
+
+	// HelpText is visible in the UI next to the filed
+	HelpText string `json:"helpText,omitempty"`
+
+	// Required indicates if the control has to be set
+	Required bool `json:"required,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=helm
+type TemplateMethod string
+
+type ApplicationTemplate struct {
+	Source   ApplicationSource `json:"source"`
+	Method   TemplateMethod    `json:"method"`
+	FormSpec []FormField       `json:"formSpec,omitempty"`
+}
+
+type VersionConstraints string
+
+func (c VersionConstraints) SemverConstraints() (*semverlib.Constraints, error) {
+	return semverlib.NewConstraint(string(c))
+}
+
+type ApplicationConstraints struct {
```

Comment:
I think we can offload the semVer validation directly into the CRD by using a kubebuilder marker with pattern validation. Advantage would be that the semVer is directly checked when applying a CR + users can actually see the regex in the CRD as well. 

Currently I have used the following marker. The REGEX is taken from the masterminds semVer library https://github.com/Masterminds/semver/blob/master/version.go#L42.

`+kubebuilder:validation:Pattern:=v?([0-9]+)(\.[0-9]+)?(\.[0-9]+)?(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?`

Reply:
@SimonTheLeg  In fact, the content of the field is not a version, but a  `semverlib.Constraints`. it can be (list is not exhaustive):

- a version: `1.2.3`
- a version with wildcard `1.18.*`
- a range of version `>= 1.16.0, < 1.19.15`

AFAIK there is no regex for checking constraints, even so, I doubt that it will be easily understandable...
The SemverConstraints has a validate Validate method that checks if a version comply with the constraint

Moreover, we will have to test that the application is "compatible" with the user cluster in several places:
- in ApplicationInstallerController to not install "incompatible" app
- in KPP API to return a list of compatible applications to the user when he installs an app through the dashboard.
I think validation logic should be in one place and close to the data to validate.


Concerning the crd:

- I have updated the description with these examples to avoid such confusion
- i think the validation should be done in the validating webhook

Reply:
Ah that's right! Yeah I guess then in your case this particular REGEX does not work. In my case for the AppRef it works (unless we want the version ref to allow ranges, which I am not in be in favor of). 

> AFAIK there is no regex for checking constraints, even so, I doubt that it will be easily understandable...
The SemverConstraints has a validate Validate method that checks if a version comply with the constraint

Well there is a REGEX which the Validate func uses ;) (https://github.com/Masterminds/semver/blob/49c09bfed6adcffa16482ddc5e5588cffff9883a/constraints.go#L147). But yes overall I agree: If the validation becomes too complicated let's have the webhook do it.

> Moreover, we will have to test that the application is "compatible" with the user cluster in several places:

I did not notice this before, but since you mentioned it: Couldn't we set the type of the field to `semverlib.Constraints` and then build a custom Unmarshaller? Then later you would never have to do any checks or func calls for conversions downstream anymore. Of course for this to work you need to have some sort of pre-apply checking.

Reply:
Okay I have looked into this since Vince is on well-deserved holidays and it turns out: In the semVer lib there is already an unmarshaller for the version type but not yet for constraints. The [original issue](https://github.com/Masterminds/semver/issues/130#issuecomment-994812618) was closed without any conclusion.  However I think it makes a lot more sense to contribute to semVer directly without wrapping the type and implement a custom marshaller in our code. Additionally, I think it does not make a difference for the end-user what type we use for this field, as they are always just going to input a string. With this I think we can leave in the string for now and always change it later on ✅